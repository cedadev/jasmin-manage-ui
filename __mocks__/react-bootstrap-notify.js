export const useNotifications = jest.fn(() => jest.fn());