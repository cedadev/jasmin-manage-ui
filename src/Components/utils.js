import React, { useEffect, useRef, useState } from 'react';

import Col from 'react-bootstrap/Col';
import Row from 'react-bootstrap/Row';
import Spinner from 'react-bootstrap/Spinner';
import Tabs from 'react-bootstrap/Tabs';
import Tab from 'react-bootstrap/Tab';

import ReactSelect from 'react-select';

import ReactMarkdown from 'react-markdown';

import '../css/form-controls.css';


/**
 * Sort the data using the values produced by the key function.
 */
export const sortByKey = (data, keyFn) => [...data].sort(
    (e1, e2) => {
        const [k1, k2] = [keyFn(e1), keyFn(e2)];
        return (k1 > k2 ? 1 : (k1 < k2 ? -1 : 0));
    }
);


export const SpinnerWithText = ({ children, justify = 'start', ...spinnerProps }) => (
    <div className={`d-flex align-items-center justify-content-${justify}`}>
        <Spinner className="mr-2" animation="border" {...spinnerProps} />
        <span>{children}</span>
    </div>
);


export const notificationFromError = error => {
    let title, message;
    if( error.name === "HttpError" ) {
        title = error.statusText;
        // Extract the most useful information from the error for the message
        // For 4xx errors, the server should always return valid JSON
        // For 5xx errors, all bets are off so just use the response text
        if( error.status >= 400 && error.status < 500 ) {
            const errorObj = error.json();
            message = errorObj.detail || JSON.stringify(errorObj);
        }
        else {
            message = error.responseText;
        }
    }
    else {
        // For any other error, just use the name and message of the error
        title = error.name;
        message = error.message;
    }
    return { level: 'danger', title, message };
};


/**
 * Wrapper for a form control that calls the onChange handler with the new value
 * instead of an event.
 */
export const Control = ({
    // The change handler, which will receive the value
    onChange,
    // The component that will be rendered
    as: Component,
    // This function extracts a value from the change event for the component
    // It receives the arguments that are passed to the onChange handler
    valueFromChangeEvent,
    ...props
}) => {
    const handleChange = (...args) => onChange(valueFromChangeEvent(...args));
    return <Component {...props} onChange={handleChange} />;
};


/**
 * Wrapper for an input that calls the onChange handler with the new value
 * instead of an event.
 */
export const Input = props => {
    const extractValue = event => event.target.value;
    return <Control {...props} as="input" valueFromChangeEvent={extractValue} />;
};


/**
 * Wrapper for a textarea that calls the onChange handler with the new value
 * instead of an event.
 */
export const TextArea = props => {
    const extractValue = event => event.target.value;
    return <Control {...props} as="textarea" valueFromChangeEvent={extractValue} />;
};


/**
 * Provides a markdown editor that can be used as a Bootstrap form control.
 */
export const MarkdownEditor = ({ value, ...props }) => {
    const [selectedTab, setSelectedTab] = useState("write");
    return (
        <div className="markdown-editor">
            <Tabs activeKey={selectedTab} onSelect={setSelectedTab}>
                <Tab eventKey="write" title="Write">
                    <TextArea value={value} rows={5} {...props} />
                    <small className="form-text text-muted">
                        This field may contain <a href="https://www.markdownguide.org/" target="_blank">Markdown</a>.
                    </small>
                </Tab>
                <Tab eventKey="preview" title="Preview">
                    <ReactMarkdown source={value} />
                </Tab>
            </Tabs>
        </div>
    );
};


const noop = () => {/* NOOP */};


/**
 * Wrapper for react-select that lets it behave a bit more like a normal select.
 */
export const Select = ({
    options,
    value,
    onChange,
    disabled,
    required,
    getOptionLabel = option => option.label,
    getOptionValue = option => option.value,
    className = '',
    ...props
}) => {
    // Store the current value as internal state
    const [state, setState] = useState(value || '');
    // When the value changes, use it to set the state
    useEffect(() => { setState(value); }, [value]);
    // Maintain a reference to the select that we will use to correctly maintain focus
    const selectRef = useRef(null);
    // When the select is changed, update the internal state and call the handler
    const handleSelectChange = option => {
        const optionValue = getOptionValue(option);
        setState(optionValue);
        onChange(optionValue);
    };
    // Sort the options by the label
    const sortedOptions = sortByKey(options, getOptionLabel);
    // Select the option that corresponds to the given value
    const selectedOption = sortedOptions.find(opt => getOptionValue(opt) === value);
    // Render the select with a hidden text input that implements required
    return (
        <div className={`${className} rich-select`}>
            <ReactSelect
                {...props}
                options={sortedOptions}
                value={selectedOption}
                onChange={handleSelectChange}
                ref={selectRef}
                isDisabled={disabled}
                getOptionLabel={getOptionLabel}
                getOptionValue={getOptionValue}
            />
            <input
                tabIndex={-1}
                autoComplete="off"
                style={{
                    opacity: 0,
                    width: "100%",
                    height: 0,
                    position: "absolute"
                }}
                value={state}
                onChange={noop}
                // When the text input is focussed as part of the required validation,
                // pass the focus onto the select
                onFocus={() => selectRef.current.focus()}
                required={required}
                disabled={disabled}
            />
        </div>
    );
};
