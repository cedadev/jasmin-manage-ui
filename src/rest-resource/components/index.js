// Export the delete button and status components directly
export * from './InstanceDeleteButton';
export * from './Status';

// Export the form pieces in their own namespace
import * as Form from './ResourceForm';
export { Form };
