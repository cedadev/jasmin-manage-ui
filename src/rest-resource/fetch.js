import Cookies from 'js-cookie';


class HttpError extends Error {
    constructor(status, statusText, responseText) {
        super(statusText);
        this.name = this.constructor.name;
        this.status = status;
        this.statusText = statusText;
        this.responseText = responseText;
    }

    json() {
        return JSON.parse(this.responseText)
    }
}


// Utility function that asynchronously waits for the specified duration
const sleep = (ms) => (new Promise(resolve => setTimeout(resolve, ms)));


const apiFetch = async (url, {
    method = 'GET',
    headers = {},
    data,
    body,
    ...options
} = {}) => {
    // Populate the required headers for the request
    const defaultHeaders = {};
    // For POST/PUT/DELETE, declare the content type and include the CSRF token if present
    if( ['POST', 'PUT', 'DELETE'].includes(method.toUpperCase()) ) {
        defaultHeaders['Content-Type'] = 'application/json';
        const csrfToken = Cookies.get('csrftoken');
        if( csrfToken ) defaultHeaders['X-CSRFToken'] = csrfToken;
    }
    // Make the actual request, injecting the cookie credentials and headers
    const response = await fetch(url, {
        method,
        headers: { ...defaultHeaders, ...headers },
        // If data is given, encode it as JSON and use it as the request body
        // If not, use the given body
        body: data ? JSON.stringify(data) : body,
        // Always include credentials
        credentials: 'include',
        // Forward any other options that were supplied
        ...options
    });
    // For a 204, just return
    if( response.status === 204 ) return;
    // Any other successful response should be JSON
    if( response.ok ) return await response.json();
    // An error response may not be JSON, so read the response as text
    const responseText = await response.text();
    throw new HttpError(response.status, response.statusText, responseText);
};


// Export method-specific versions of apiFetch
export default apiFetch;
