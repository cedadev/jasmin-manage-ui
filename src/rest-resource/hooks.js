import { useCallback, useEffect, useState } from 'react';

import apiFetch from './fetch';


/**
 * Hook that fetches data from an HTTP endpoint and passes it to a callback.
 */
export const useOnFetch = (url, { onFetch, autoFetch = true } = {}) => {
    // Indicates whether a fetch is currently in progress
    // Setting this to true initially will cause the data to be loaded
    const [fetching, setFetching] = useState(autoFetch);
    // The error from the last fetch if it was not successful
    const [fetchError, setFetchError] = useState(null);

    // When a fetch is triggered by changing the state, fetch the data
    // Doing it this way ensures that we never have two fetches competing, as the fetch
    // is only triggered when fetching changes from false to true
    useEffect(
        () => {
            if( !fetching ) return;
            // useEffect expects the return value to be a cleanup function
            // This means we can't pass an async function directly as that returns a promise
            // Hence we have to make and call an anonymous function
            (async () => {
                try {
                    // Fetch the data from the endpoint
                    const result = await apiFetch(url);
                    // Update the data
                    onFetch(result);
                    // Clear any stored error
                    setFetchError(null);
                }
                catch(error) {
                    // For an error response, store the response
                    setFetchError(error);
                }
                finally {
                    // In all cases, the fetch is over
                    setFetching(false);
                }
            })();
        },
        [fetching]
    );

    // This function triggers a fetch by updating the state
    const fetch = useCallback(() => setFetching(true), []);

    // Return the public state for consumers to use
    return { fetching, fetchError, fetch };
};


/**
 * Hook for fetching data from an HTTP endpoint.
 */
export const useEndpoint = (url, { initialData, autoFetch = true } = {}) => {
    // Indicates if the endpoint has been successfully fetched at least once
    const [initialised, setInitialised] = useState(initialData !== undefined);
    // Make a variable to store the data and use it to initialise the hook
    const [data, setData] = useState(initialData);
    // When there is a successful fetch, set data and initialised
    const onFetch = data => { setData(data); setInitialised(true); };
    // Return the public state for consumers to use
    return { ...useOnFetch(url, { onFetch, autoFetch }), initialised, data };
};


/**
 * Hook for running a mutation against an HTTP endpoint.
 */
export const useMutation = (url, method) => {
    const [inProgress, setInProgress] = useState(false);
    // This function runs the mutation
    const mutate = async data => {
        setInProgress(true);
        try {
            // Attempt the mutation
            return await apiFetch(url, { method, data });
        }
        finally {
            // In all cases, the mutation is over
            setInProgress(false);
        }
    };
    // Return the public state for consumers to use
    return { inProgress, mutate };
};


/**
 * Hook for using an endpoint as a REST resource.
 */
export const useResource = (resourceUrl, { initialData, autoFetch = true } = {}) => {
    // Indicates if the resource has been successfully fetched at least once
    const [initialised, setInitialised] = useState(initialData !== undefined);
    // We store the current data as an object mapping id => instance
    const [data, setData] = useState(initialData || {});

    // This function updates the data of a single instance in the mapping
    const setInstanceData = id => transform => setData(state => {
        const nextInstanceState = transform instanceof Function ?
            transform(state[id]) :
            transform;
        // Allow the instance to set state to undefined to remove itself
        if( nextInstanceState !== undefined ) {
            return { ...state, [id]: nextInstanceState };
        }
        else {
            const { [id]: _, ...nextState } = state;
            return nextState;
        }
    });

    // When there is a successful fetch, set data and initialised
    // The data will be received in list form, which we transform into the object format
    const onFetch = instanceList => {
        setData(
            Object.assign(
                {},
                ...instanceList.map(instanceData => ({
                    [instanceData.id]: newInstance(
                        instanceData,
                        setInstanceData(instanceData.id)
                    )
                }))
            )
        )
        setInitialised(true);
    };

    // Initialise the on-fetch hook that will be responsible for fetching
    // the resource list
    const fetchState = useOnFetch(resourceUrl, { onFetch, autoFetch });

    // Use a mutation to create new instances and add them to the data
    const { inProgress: creating, mutate } = useMutation(resourceUrl, 'POST');
    const create = async data => {
        const instanceData = await mutate(data);
        setData(state => ({
            ...state,
            [instanceData.id]: newInstance(
                instanceData,
                setInstanceData(instanceData.id)
            )
        }));
    };

    // Return the pieces that comprise the public interface
    return { ...fetchState, data, initialised, creating, create };
};


/**
 * Hook for using a single resource instance independently of the resource.
 */
export const useResourceInstance = (instanceUrl, { initialData, autoFetch = true } = {}) => {
    // Indicates if the instance has been successfully fetched at least once
    const [initialised, setInitialised] = useState(initialData !== undefined);
    // We store the current data as an object mapping id => instance
    const [data, setData] = useState(initialData);
    // When there is a successful fetch, set data and initialised
    const onFetch = data => { setData(data); setInitialised(true); };
    // Initialise the fetch state
    const fetchState = useOnFetch(instanceUrl, { onFetch, autoFetch });

    // Use mutations to update and delete the resource, updating the data as required
    const { inProgress: updating, mutateUpdate } = useMutation(instanceUrl, 'PUT');
    const update = async data => setData(await mutateUpdate(data));

    const { inProgress: deleting, mutateDelete } = useMutation(instanceUrl, 'DELETE');
    const delete_ = async () => setData(await mutateDelete());

    // Executing an action is slightly more difficult as the URL depends on the action
    const [actionInProgress, setActionInProgress] = useState(false);
    const executeAction = async (name, actionData) => {
        setActionInProgress(true);
        try {
            // Attempt to apply the action
            const instanceData = await apiFetch(
                data._links[name],
                { method: 'POST', data: actionData }
            );
            setData(instanceData);
        }
        finally {
            setActionInProgress(false);
        }
    };

    // Return the pieces that comprise the public interface
    return {
        ...fetchState,
        data,
        initialised,
        updating: updating || actionInProgress,
        update,
        deleting,
        delete: delete_,
        executeAction
    };
};


// The setState function received by newInstance is scoped to update the data for the resource
const newInstance = (initialData, setState) => ({
    // Indicates if the instance is currently being fetched
    fetching: false,
    // Indicates if the instance is currently being updated
    updating: false,
    // Indicates if the instance is currently being deleted
    deleting: false,
    // The error from the last fetch if it was not successful
    fetchError: null,
    // The current data for the instance
    data: initialData,
    // Re-fetch the data for the instance
    fetch: async () => {
        // Mark the fetch as in progress
        setState(state => ({ ...state, fetching: true }));
        try {
            // Attempt to fetch the instance
            const instanceData = await apiFetch(initialData._links.self);
            // Update the state if successful
            setState(state => ({ ...state, fetching: false, data: instanceData }));
        }
        catch(error) {
            // If the fetch is not successful, store the error
            setState(state => ({ ...state, fetching: false, fetchError: error }));
        }
    },
    // Update the instance with new data
    update: async data => {
        // Make the update as in progress
        setState(state => ({ ...state, updating: true }));
        try {
            // Attempt to update the instance
            const instanceData = await apiFetch(initialData._links.self, { method: 'PUT', data });
            // Update the state if the update was successful
            setState(state => ({ ...state, updating: false, data: instanceData }));
        }
        catch(error) {
            // On failure, mark the update as over before re-throwing the error
            setState(state => ({ ...state, updating: false }));
            throw error;
        }
    },
    // Delete the instance
    delete: async () => {
        // Mark the delete as in progress
        setState(state => ({ ...state, deleting: true }));
        try {
            // Attempt to delete the instance
            await apiFetch(initialData._links.self, { method: 'DELETE' });
            // Remove the entry if the delete is successful
            setState(undefined);
        }
        catch(error) {
            // On failure, mark the delete as over before re-throwing the error
            setState(state => ({ ...state, deleting: false }));
            throw error;
        }
    },
    // Executes a named action on the instance
    executeAction: async (name, data) => {
        // Mark the instance as updating
        setState(state => ({ ...state, updating: true }));
        try {
            // Attempt to apply the action
            const instanceData = await apiFetch(initialData._links[name], { method: 'POST', data });
            // Update the state if the action was successful
            setState(state => ({ ...state, updating: false, data: instanceData }));
        }
        catch(error) {
            // On failure, mark the update as over before re-throwing the error
            setState(state => ({ ...state, updating: false }));
            throw error;
        }
    }
});
