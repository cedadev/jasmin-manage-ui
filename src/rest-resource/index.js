export * from './components';
export * from './context';
export * from './hooks';

import apiFetch from './fetch';
export { apiFetch };
