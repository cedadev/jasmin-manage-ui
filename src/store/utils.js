import produce from 'immer';


// Maps the values of an object, maintaining the keys
// The map function receives the value as the first argument and the key as the second
// and should return a new value
export const mapValues = (mapFn, obj) => Object.fromEntries(
    Object.entries(obj).map(([k, v]) => [k, mapFn(v, k)])
);


// Behaves like Python's dict.setdefault
export const setDefault = (obj, key, defaultValue) => {
    // First, make sure that the key exists in the object
    if( !obj.hasOwnProperty(key) ) obj[key] = defaultValue;
    // Then return the value of the key
    return obj[key];
}


// Decorator that replaces zustand's "set" with a "mutate" function
// that allows state to be mutated using immer
export const immer = config => (set, get, api) => config(fn => set(produce(fn)), get, api);


// Take existing mutate and getState functions and a key and returns new mutate and
// getState functions that operate on only that subset of the state
// The state modification function passed to the scoped mutate can optionally receive the
// parent state as a second argument if required, e.g. to remove itself
// We also return a third function that allows access to the parent state
export const withScope = (mutate, getState, scope) => {
    // The scope is a single key or a list of keys
    const keys = Array.isArray(scope) ? scope : [scope];
    // To get the scoped state, descend into the object with the given keys
    const scopedState = state => keys.reduce((s, k) => setDefault(s, k, {}), state);
    return [
        stateFn => mutate(state => { stateFn(scopedState(state), state); }),
        () => scopedState(getState()),
        getState
    ];
};
