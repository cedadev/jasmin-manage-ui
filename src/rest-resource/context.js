import React, { createContext, useContext, useEffect } from 'react';

import { useEndpoint, useResource, useInstance } from './hooks';


// This higher order function turns a hook into a context, allowing the hook
// state to be shared between components
const createContextForHook = hookFunction => (url, options = {}) => {
    // Make a new context
    const Context = createContext();
    // Return a provider and hook that use the context to share the hook state
    return {
        // The provider is an anchor for the data and also a fetch point
        // However, we do not auto-fetch the data as soon as the provider as mounted
        Provider: ({ children }) => {
            const fetchable = hookFunction(url, { ...options, autoFetch: false });
            return <Context.Provider value={fetchable}>{children}</Context.Provider>;
        },
        // The first time a component using the hook is mounted, the data for the fetchable
        // is initialised if it has not already been initialised
        hook: (autoFetch = true) => {
            const fetchable = useContext(Context);
            useEffect(
                // If the fetchable is not initialised, trigger a fetch by marking it as dirty
                // This should not trigger multiple fetches
                () => { if( autoFetch && !fetchable.initialised ) fetchable.markDirty(); },
                []
            );
            return fetchable;
        }
    };
};


export const createContextForEndpoint = createContextForHook(useEndpoint);
export const createContextForResource = createContextForHook(useResource);
export const createContextForInstance = createContextForHook(useInstance);
