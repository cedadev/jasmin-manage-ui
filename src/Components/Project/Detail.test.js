// src/Components/Project/Detail.test.js

import React from "react";
import { render, screen, waitFor, fireEvent } from "@testing-library/react";
import { MemoryRouter, Route } from "react-router-dom";
import renderer from "react-test-renderer";
import userEvent from "@testing-library/user-event";

import ProjectDetailWrapper from "./Detail";
import { useNotifications } from "react-bootstrap-notify";
import {
  useProject,
  useProjectEvents,
  useCurrentUser,
  useConsortia,
  useCategories,
  useResources,
} from "../../api";
import { useNestedResource } from "../../rest-resource";
import { notificationFromError } from "../utils";

// Mock the API hooks
jest.mock("../../api", () => ({
  useProject: jest.fn(),
  useProjectEvents: jest.fn(),
  useCurrentUser: jest.fn(),
  useConsortia: jest.fn(),
  useCategories: jest.fn(),
  useResources: jest.fn(),
}));

// Define MockStatus and its subcomponents
const MockStatus = ({ children }) => <div>{children}</div>;

MockStatus.Loading = ({ children }) => <div>{children}</div>;
MockStatus.Available = ({ children }) => <div>{children}</div>;
MockStatus.Unavailable = ({ children }) => <div>{children}</div>;

// Mock the rest-resource hooks including Status
jest.mock("../../rest-resource", () => ({
  ...jest.requireActual("../../rest-resource"),
  useNestedResource: jest.fn(),
}));

// Mock the notifications
jest.mock("react-bootstrap-notify", () => ({
  useNotifications: jest.fn(),
}));

const mockProject = {
  data: {
    id: 1,
    name: "Project A",
    description: "Description of Project A",
    status: "ACTIVE",
    consortium: 1,
  },
  fetchError: null,
  fetching: false,
  initialised: true,
};

const mockConsortia = {
  data: {
    1: {
      data: {
        id: 1,
        name: "Consortium A",
        manager: { id: 1 },
      },
    },
  },
  fetchError: null,
  fetching: false,
  initialised: true,
};

const mockCurrentUser = {
  data: {
    id: 1,
    username: "testuser",
    first_name: "Test",
    last_name: "User",
    email: "test@example.com",
  },
  fetchError: null,
  fetching: false,
  initialised: true,
};

const mockNotifications = jest.fn();

describe("Testing Project Detail Page", () => {
  beforeEach(() => {
    useCurrentUser.mockReturnValue(mockCurrentUser);
    useConsortia.mockReturnValue(mockConsortia);
    useProject.mockReturnValue(mockProject);
    useProjectEvents.mockReturnValue([]);
    useCategories.mockReturnValue({
      data: {},
      fetchError: null,
      fetching: false,
      initialised: true,
    });
    useNestedResource.mockReturnValue({
      data: {},
      fetchError: null,
      fetching: false,
      initialised: true,
    });
    useNotifications.mockReturnValue(mockNotifications);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("Should match the snapshot", () => {
    const component = renderer.create(
      <MemoryRouter initialEntries={["/projects/1"]}>
        <Route path="/projects/:id">
          <ProjectDetailWrapper />
        </Route>
      </MemoryRouter>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Should display the project name", async () => {
    render(
      <MemoryRouter initialEntries={["/projects/1"]}>
        <Route path="/projects/:id">
          <ProjectDetailWrapper />
        </Route>
      </MemoryRouter>
    );

    await waitFor(() => {
      expect(screen.getByText("Project A")).toBeInTheDocument();
    });
  });

  it("Should display the 'Loading project...' text when fetching", () => {
    useProject.mockReturnValue({
      data: null,
      fetchError: null,
      fetching: true,
      initialised: false,
    });

    render(
      <MemoryRouter initialEntries={["/projects/1"]}>
        <Route path="/projects/:id">
          <ProjectDetailWrapper />
        </Route>
      </MemoryRouter>
    );

    expect(screen.getByText("Loading project...")).toBeInTheDocument();
  });

  it("Should redirect to /projects when project is unavailable", async () => {
    useProject.mockReturnValue({
      data: { ...mockProject.data },
    });

    render(
      <MemoryRouter initialEntries={["/projects/1"]}>
        <Route path="/projects/:id">
          <ProjectDetailWrapper />
        </Route>
        <Route path="/projects">
          <div>Projects List Page</div>
        </Route>
      </MemoryRouter>
    );

    await waitFor(() => {
      expect(screen.getByText("Projects List Page")).toBeInTheDocument();
    });
  });

  it("Should notify the user when project fetch fails", async () => {
    useProject.mockReturnValue({
      data: { ...mockProject.data },
      fetchError: new Error("Failed to fetch project"),
      fetching: false,
      initialised: true,
    });

    render(
      <MemoryRouter initialEntries={["/projects/1"]}>
        <Route path="/projects/:id">
          <ProjectDetailWrapper />
        </Route>
      </MemoryRouter>
    );

    await waitFor(() => {
      expect(mockNotifications).toHaveBeenCalledWith(
        notificationFromError(new Error("Failed to fetch project"))
      );
    });
  });

  it("Should display consortium manager instructions when project is under review and user is manager", async () => {
    useProject.mockReturnValue({
      ...mockProject,
      data: { ...mockProject.data, status: "UNDER_REVIEW" },
    });

    useConsortia.mockReturnValueOnce({
      ...mockConsortia,
      data: {
        1: {
          data: {
            ...mockConsortia.data[1].data,
            manager: { id: 1 }, // Current user is manager
          },
        },
      },
    });

    render(
      <MemoryRouter initialEntries={["/projects/1"]}>
        <Route path="/projects/:id">
          <ProjectDetailWrapper />
        </Route>
      </MemoryRouter>
    );

    const servicesTab = screen.getByRole("link", { name: "Services" });
    fireEvent.click(servicesTab);

    await waitFor(() => {
      expect(screen.getByText(/please click either/i)).toBeInTheDocument();
      expect(
        screen.getByText(
          /once you have made decisions on all the requirements for the project./i
        )
      ).toBeInTheDocument();
    });
  });

  it("Should display edit instructions when project is editable and user is not manager", async () => {
    useProject.mockReturnValue({
      ...mockProject,
      data: { ...mockProject.data, status: "EDITABLE" },
    });

    // Mock current user as ID 2 (non-manager)
    useCurrentUser.mockReturnValue({
      data: {
        id: 2,
        username: "testuser2",
        first_name: "Test",
        last_name: "User2",
        email: "test2@example.com",
      },
      fetchError: null,
      fetching: false,
      initialised: true,
    });

    // Keep manager as ID 1
    useConsortia.mockReturnValue({
      ...mockConsortia,
      data: {
        1: {
          data: {
            ...mockConsortia.data[1].data,
            manager: { id: 1 },
          },
        },
      },
    });

    render(
      <MemoryRouter initialEntries={["/projects/1"]}>
        <Route path="/projects/:id">
          <ProjectDetailWrapper />
        </Route>
      </MemoryRouter>
    );

    const servicesTab = screen.getByRole("link", { name: "Services" });
    fireEvent.click(servicesTab);

    await waitFor(() => {
      expect(
        screen.getByText(/Please ensure that you click the/i)
      ).toBeInTheDocument();
      expect(
        screen.getByText(
          /button once you have added your requirements to all services./i
        )
      ).toBeInTheDocument();
    });
  });

  it("Should display delete/edit instructions when there are rejected requirements and user is not manager", async () => {
    // Mock resources
    useResources.mockReturnValue({
      data: {
        1: {
          data: {
            id: 1,
            name: "Resource 1",
            short_name: "R1",
            units: "GB",
          },
        },
      },
      initialised: true,
    });

    // Mock categories
    useCategories.mockReturnValue({
      data: {
        1: {
          data: {
            id: 1,
            name: "Category 1",
            resources: [1],
          },
        },
      },
      initialised: true,
    });

    // Mock project with rejected requirements
    useNestedResource.mockImplementation((resource, name) => {
      const mockInstance = {
        data: {
          1: {
            data: {
              id: 1,
              name: "Service 1",
              category: 1,
              _links: {
                requirements: "/api/services/1/requirements/",
              },
            },
            nestedResourceMethods: (name, url, options) => ({
              fetch: () => {},
              markDirty: () => {},
              reset: () => {},
            }),
            nestedResources: {
              requirements: {
                data: {
                  1: {
                    data: {
                      id: 1,
                      status: "REJECTED",
                      resource: 1,
                      amount: 100,
                    },
                  },
                },
                initialised: true,
              },
            },
          },
        },
        initialised: true,
        nestedResourceMethods: (name, url, options) => ({
          fetch: () => {},
          markDirty: () => {},
          reset: () => {},
        }),
      };

      if (name === "services") {
        return mockInstance;
      }
      if (name === "requirements") {
        return {
          data: {
            1: {
              data: {
                id: 1,
                status: "REJECTED",
                resource: 1,
                amount: 100,
              },
            },
          },
          initialised: true,
        };
      }
      return { data: {}, initialised: true };
    });

    // Mock current user as non-manager (ID 2)
    useCurrentUser.mockReturnValue({
      data: {
        id: 2,
        username: "testuser2",
        first_name: "Test",
        last_name: "User2",
        email: "test2@example.com",
      },
      fetchError: null,
      fetching: false,
      initialised: true,
    });

    // Keep consortium manager as ID 1
    useConsortia.mockReturnValue({
      ...mockConsortia,
      data: {
        1: {
          data: {
            ...mockConsortia.data[1].data,
            manager: { id: 1 },
          },
        },
      },
    });

    render(
      <MemoryRouter initialEntries={["/projects/1"]}>
        <Route path="/projects/:id">
          <ProjectDetailWrapper />
        </Route>
      </MemoryRouter>
    );

    const servicesTab = screen.getByRole("link", { name: "Services" });
    fireEvent.click(servicesTab);

    await waitFor(() => {
      expect(screen.getByText(/Please/i)).toBeInTheDocument();
      expect(
        screen.getByText(
          /any rejected requirements before submitting your project for review./i
        )
      ).toBeInTheDocument();
    });
  });
});
