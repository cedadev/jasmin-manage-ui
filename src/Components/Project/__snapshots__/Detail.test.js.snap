// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Project Detail Page Should match the snapshot 1`] = `
[
  <div
    className="row"
  >
    <div
      className="col"
    >
      <h1
        className="border-bottom mt-4 mb-3"
      >
        Project A
      </h1>
    </div>
  </div>,
  "",
  <div
    className="row"
  >
    <div
      className="order-lg-1 col-xxl-3 col-xl-4 col-lg-5 col-12"
    >
      <div
        className="sticky-top pt-3"
      >
        <div
          className="card"
          style={
            {
              "borderWidth": "2px",
            }
          }
        >
          <div
            className="list-group list-group-flush"
            onKeyDown={[Function]}
          >
            <div
              className="list-group-item"
              data-rb-event-key={null}
              disabled={false}
              onClick={[Function]}
            >
              Project is 
              <strong>
                ACTIVE
              </strong>
              .
            </div>
            <div
              className="list-group-item"
              data-rb-event-key={null}
              disabled={false}
              onClick={[Function]}
            >
              <p
                className="mb-0"
              >
                Project belongs to
                 
                <a
                  className="font-weight-bold"
                  href="/consortia/1"
                  onClick={[Function]}
                >
                  Consortium A
                </a>
                .
              </p>
              <p
                className="mt-2 mb-0"
              >
                <strong>
                  You are the consortium manager.
                </strong>
              </p>
            </div>
            <div
              className="list-group-item"
              data-rb-event-key={null}
              disabled={false}
              onClick={[Function]}
            >
              <p
                className="mb-0"
              >
                Project has
                 
                <button
                  className="p-0 btn btn-link"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  <strong>
                    0
                     collaborator
                    s
                  </strong>
                </button>
                .
              </p>
            </div>
            <div
              className="list-group-item"
              data-rb-event-key={null}
              disabled={false}
              onClick={[Function]}
            >
              Project was created 
              <strong>
                a few seconds ago
              </strong>
              .
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="order-lg-0 col-xxl-9 my-3 col-xl-8 col-lg-7 col-12"
    >
      <div
        className="mb-3 nav nav-tabs"
        onKeyDown={[Function]}
      >
        <div
          className="nav-item"
        >
          <a
            className="active nav-link active"
            data-rb-event-key="/projects/1"
            href="/projects/1"
            onClick={[Function]}
            onKeyDown={[Function]}
            style={{}}
          >
            Overview
          </a>
        </div>
        <div
          className="nav-item"
        >
          <a
            className="nav-link"
            data-rb-event-key="/projects/1/services"
            href="/projects/1/services"
            onClick={[Function]}
            onKeyDown={[Function]}
          >
            Services
          </a>
        </div>
      </div>
      <div
        className="timeline"
      >
        <div
          className="timeline-item"
        >
          <div
            className="timeline-item-container"
          >
            <div
              className="timeline-item-content"
            >
              <div
                className="card"
              >
                <div
                  className="card-header"
                >
                  <strong>
                    Project description
                  </strong>
                </div>
                <div
                  className="markdown card-body"
                >
                  <p>
                    Description of Project A
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
